Add JIT support for cmath unary ops

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54692 Add support for real and imag tensor attributes
* #54541 [TS] Add complex support for more ops
* #54328 Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants
* **#54089 Add JIT support for cmath unary ops**

**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

Differential Revision: [D27339149](https://our.internmc.facebook.com/intern/diff/D27339149)

anjali411 committed:
Add JIT support for cmath unary ops

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"




[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"




[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"




[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"




[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"




[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"




[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


This PR adds:
1. support for the following cmath functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase`

Future work:
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


This PR adds:
1. support for the following cmath functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase`

Future work:
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


This PR adds:
1. support for the following cmath functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase`

Future work:
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

SplitInfinity reviewed (commented):
**This comment was meant for another PR but cannot be deleted for some reason so I have replaced it with this text**

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,75 @@ def fn(a: int, b: torch.Tensor, dim: int):
+
+        def checkMath(func_name, num_args, ret_type="complex", debug=False, vals=None, args_type=None):
+            funcs_template = dedent('''
+            def func(a):
+                # type: {args_type} -> {ret_type}
```
```suggestion
            def func(a: complex) -> {ret_type}:
```
Python 2 is no longer supported.

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,75 @@ def fn(a: int, b: torch.Tensor, dim: int):
+
+    def test_complex_math_ops(self):
+        inf = float("inf")
+        NaN = float("nan")
+        vals = ([inf, NaN, 0.0, 1.0, 2.2, -1.0, -0.0, -2.2, -inf, 1, 0, 2]
```
```suggestion
        vals = ([inf, float("nan"), 0.0, 1.0, 2.2, -1.0, -0.0, -2.2, -inf, 1, 0, 2]
```

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,75 @@ def fn(a: int, b: torch.Tensor, dim: int):
+                    res_script = f_script(a)
+                except Exception as e:
+                    res_script = e
+                if debug:
+                    print("in: ", a)
+                    print("out: ", res_python, res_script)
```
```suggestion
```
Debug code?

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,75 @@ def fn(a: int, b: torch.Tensor, dim: int):
+                + [10.0 ** i for i in range(5)] + [-(10.0 ** i) for i in range(5)])
+        complex_vals = tuple(complex(x, y) for x, y in product(vals, vals))
+
+        def checkMath(func_name, num_args, ret_type="complex", debug=False, vals=None, args_type=None):
```
```suggestion
        def checkMath(func_name, num_args, ret_type="complex"):
```

SplitInfinity replied:
`num_args` seems to be 1 in every invocation, so it seems like that doesn't need to be a parameter either.

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,75 @@ def fn(a: int, b: torch.Tensor, dim: int):
+                args = "a, b"
+            else:
+                raise RuntimeError("Test doesn't support more than 2 arguments")
+            if args_type is None:
+                args_type = "(complex)"
```
```suggestion
```
Argument does not seem to be used.

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,75 @@ def fn(a: int, b: torch.Tensor, dim: int):
+                raise RuntimeError("Test doesn't support more than 2 arguments")
+            if args_type is None:
+                args_type = "(complex)"
+            funcs_str = funcs_template.format(func=func_name, args=args, args_type=args_type, ret_type=ret_type)
```
```suggestion
            funcs_str = funcs_template.format(func=func_name, args=args, ret_type=ret_type)
```

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,75 @@ def fn(a: int, b: torch.Tensor, dim: int):
+                           .format(func_name=func_name, a=a, res_python=res_python, res_script=res_script))
+                    self.assertEqual(res_python, res_script, msg=msg)
+
+        unary_ops = ['log', 'log10', 'sqrt', 'exp', 'sin', 'cos', 'asin', 'acos', 'atan', 'sinh', 'cosh',
+                     'tanh', 'asinh', 'acosh', 'atanh']
```
Is there any way to generate this list based on the contents of `cmath` so that new ops are automatically tested?

anjali411 replied:
Mmm I am not sure, but I think it's ok to stay with this list since there aren't any unary ops not present in cmath and are supported for complex in pytorch.

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,75 @@ def fn(a: int, b: torch.Tensor, dim: int):
+            funcs_template = dedent('''
+            def func(a):
+                # type: {args_type} -> {ret_type}
+                return cmath.{func}({args})
```
```suggestion
                return cmath.{func}(a)
```

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,75 @@ def fn(a: int, b: torch.Tensor, dim: int):
+                        continue
+
+                    if type(res_python) == type(res_script):
+                        if isinstance(res_python, tuple) and (cmath.isnan(res_python[0]) == cmath.isnan(res_script[0])):
+                            continue
+                        if isinstance(res_python, complex) and cmath.isnan(res_python) and cmath.isnan(res_script):
+                            continue
```
It looks like these were add to present tests with certain combinations of functions and inputs from failing. I think we should try updating `assertEqual` so that it can handle nan. My guess is that the current implementation is checking whether `a == b` and that's `False` when both are `nan`.

Also, which function returns a tuple?

anjali411 replied:
yeah that code is not needed. removed it.

SplitInfinity commented on torch/csrc/jit/runtime/register_ops_utils.h:
```diff
@@ -652,6 +652,48 @@ void listSetItem(Stack* stack);
+      },                                                                  \
+      aliasAnalysisFromSchema())
+
+#define DEFINE_UNARY_OP_WITH_COMPLEX_CAST(                                \
```
Can you add some comments explaining what this is for and why it is needed? It seems to me that it only exists to make sure that `abs` has a float result.

anjali411 replied:
yeah it seemed a waste to use this everywhere since most unary ops output complex valued result. Only a few ops return real valued result. But good idea-- I'll add a comment!

SplitInfinity reviewed (commented):
There are test failures for `exp` and `tanh` that need to be investigated.

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

malfet commented:
Posting here for posterity:
```
#include <iostream>
#include <complex>
#include <cmath>

int main(void) {
    std::complex<float> a(INFINITY, NAN);
    std::complex<float> b(-INFINITY, INFINITY);

    std::cout << std::exp(a) << std::endl;
    std::cout << std::tanh(b) << std::endl;

    return 0;
}
```
Returns
```
(inf, nan)
(-1, 0)
``` if compiled on linux by gcc
```
(inf, nan)
(1, 0)
``` if compiled on Mac by clang
```
(inf,inf)
(-nan(ind),-nan(ind))
``` if compiled on Windows by VC++

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

SplitInfinity reviewed (commented):
What was the fix for the failing tests, by the way?

anjali411 commented:
> What was the fix for the failing tests, by the way?

I removed  `inf, nan` from the values being tested since the tests are failing are due to the bugs in standard library functions on windows and mac, and does not relate to the JIT support we added in this PR.

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

SplitInfinity commented on torch/jit/_builtins.py:
```diff
@@ -55,17 +56,30 @@
     (math.asinh, "aten::asinh"),
     (math.atanh, "aten::atanh"),
     (math.acosh, "aten::acosh"),
-    (math.sinh, "aten::sinh"),
-    (math.cosh, "aten::cosh"),
-    (math.tanh, "aten::tanh"),
     (math.fmod, "aten::fmod"),
     (math.modf, "aten::modf"),
     (math.factorial, "aten::factorial"),
     (math.frexp, "aten::frexp"),
-    (math.isnan, "aten::isnan"),
```
Why have the these been deleted?

anjali411 replied:
just deleted some duplicate entries!

SplitInfinity replied:
Phew, ok. The existing entries were hidden by the preview.

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,54 @@ def fn(a: int, b: torch.Tensor, dim: int):
+                except Exception as e:
+                    res_script = e
+
+                if res_python != res_script:
+                    if isinstance(res_python, Exception):
+                        continue
```
What if Python throws an exception but TorchScript does not?

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,54 @@ def fn(a: int, b: torch.Tensor, dim: int):
+                           .format(func_name=func_name, a=a, res_python=res_python, res_script=res_script))
+                    self.assertEqual(res_python, res_script, msg=msg)
+
+        unary_ops = ['log', 'log10', 'sqrt', 'exp', 'sin', 'cos', 'asin', 'acos', 'atan', 'sinh', 'cosh',
+                     'tanh', 'asinh', 'acosh', 'atanh', 'phase']
```
I would still recommend generating these programmatically so that future ops are automatically tested but I won't not stamp because of it.

SplitInfinity reviewed (commented):
Just one more question about deleted builtin ops.

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

Differential Revision: [D27339149](https://our.internmc.facebook.com/intern/diff/D27339149)

[ghstack-poisoned]

malfet commented on test/jit/test_complex.py:
```diff
@@ -55,3 +58,54 @@ def fn(a: int, b: torch.Tensor, dim: int):
+    def test_complex_math_ops(self):
+        vals = ([0.0, 1.0, 2.2, -1.0, -0.0, -2.2, 1, 0, 2]
+                + [10.0 ** i for i in range(2)] + [-(10.0 ** i) for i in range(2)])
+        complex_vals = tuple(complex(x, y) for x, y in product(vals, vals))
```
I don't know what's the difference between Python complex literals vs complex class, but `cmath.sqrt(complex(x,y)) and `cmath.sqrt(x+y*1j)` can sometimes return different results:
```
python3 -c "import cmath;x,y=-1.0,-0.0;print(cmath.sqrt(x+y*1j), cmath.sqrt(complex(x,y)))"
1j -1j

```
```suggestion
        complex_vals = tuple(x+y*1k for x, y in product(vals, vals))
```

anjali411 replied:
It's because:
```
>>> -0.0 * 1j
(-0+0j)
```
so `x+y*1j` for `x,y=-1.0,-0.0` returns `-1+0j`. Regardless, `cmath.sqrt` returns incorrect value on `-1-0j` and we should probably file a bug for it.

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

Differential Revision: [D27339149](https://our.internmc.facebook.com/intern/diff/D27339149)

[ghstack-poisoned]

anjali411 committed:
Update on "Add JIT support for cmath unary ops"


**This PR adds:**
1. support for the following [cmath](https://docs.python.org/3/library/cmath.html) functions:
     - Power and logarithmic functions (`cmath.{exp, log, log10, sqrt}`) 
     - Trigonometric functions (`cmath.{sin, cos, tan, asin, acos, atan}`)
     - Hyperbolic functions (`cmath.{sinh, cos, tanh, asinh, acosh, atanh}`)
     - `cmath.phase()`
2. `abs()`

**Future work:**
1. support
    - `cmath.{polar, rect}` 
    - classification functions (`cmath.{isfinite, isnan, isinf, isclose}`)
    - constants (`cmath.{pi, e, inf, nan, infj, nanj}`)

Differential Revision: [D27339149](https://our.internmc.facebook.com/intern/diff/D27339149)

[ghstack-poisoned]
